File: examples/static_output.rs
use alfrusco::{config, DefaultWorkflowError, Item, Workflow};
use clap::Parser;

#[derive(Parser)]
struct StaticOutputWorkflow {}

pub fn main() {
    env_logger::init();
    let command = StaticOutputWorkflow {};
    alfrusco::execute(&config::AlfredEnvProvider, command, &mut std::io::stdout());
}

impl alfrusco::Runnable for StaticOutputWorkflow {
    type Error = DefaultWorkflowError;
    fn run(self, wf: &mut Workflow) -> Result<(), DefaultWorkflowError> {
        wf.response.skip_knowledge(true);
        wf.response.append_items(vec![
            Item::new("First Option").subtitle("First Subtitle"),
            Item::new("Option 2").subtitle("Second Subtitle"),
            Item::new("Three").subtitle("3"),
        ]);
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_static_output_workflow() {
        let command = StaticOutputWorkflow {};
        let mut buffer = Vec::new();
        let dir = tempfile::tempdir().unwrap().into_path();
        alfrusco::execute(&config::TestingProvider(dir), command, &mut buffer);
        let output = String::from_utf8(buffer).unwrap();
        assert!(output.contains("\"title\":\"First Option\""));
        assert!(output.contains("\"subtitle\":\"First Subtitle\""));
    }
}

File: haskell/examples/StaticOutput.hs
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE OverloadedStrings #-}

-- | Equivalent to Rust examples/static_output.rs
-- Outputs 3 items with skipknowledge=true
module Main (main) where

import System.IO (stdout)

import Alfrusco

data StaticOutputWorkflow = StaticOutputWorkflow

instance Runnable StaticOutputWorkflow AlfruscoError where
  run _ wf = do
    setSkipKnowledge wf True
    appendItems wf
      [ withSubtitle "First Subtitle" $ item "First Option"
      , withSubtitle "Second Subtitle" $ item "Option 2"
      , withSubtitle "3" $ item "Three"
      ]
    pure (Right ())

main :: IO ()
main = execute AlfredEnvProvider StaticOutputWorkflow stdout
